Create "Process management with PM2 guide" - Closes 1562 (#1686)
LiskHQ/lisk-docs: docs/modules/ROOT/pages/run-blockchain/process-management.adoc

= Process management with PM2
Muhammad Talha <muhammad.talha@lightcurve.io>
//Settings
:toc:
:toc: preamble

// External URLs
:url_pm2_docs: https://pm2.keymetrics.io/docs/usage/quick-start/
:url_PM2_article: https://futurestud.io/tutorials/pm2-advanced-app-configuration-with-json-file
:url_global_cli: build-blockchain/create-blockchain-app.adoc#using-the-application-cli-globally
:url_build_hello_client: build-blockchain/index.adoc#the-hello-world-client
:url_sidechain_client: glossary.adoc#sidechain-client
// footnotes
:fn_sidechain_client_glossary: footnote:client[See xref:{url_sidechain_client}[Sidechain client] for more details.]

> It is recommended to set up an additional process management tool to run a sidechain client{fn_sidechain_client_glossary}.

Reviewer: maybe we can mention to also manage multiple instances of sidechain_client

Author: Sure! For that, we'll have to use the **pm2.conf.json** file, right? Like mentioning the details of two apps in the JSON file and then running the `pm2 start pm2.conf.json` command?

Reviewer: yeah I think you can simply run a JSON file `pm2.conf.json`

Any process management tool can be used to manage the sidechain client.
However, we recommend using PM2 to manage the instance.

== Setup

Install PM2 globally:

[source,bash]
----
npm i -g pm2
----

Once PM2 is installed, you can start your sidechain client via the following two methods:

. <<using-a-bash-script>>
. <<using-a-json-config-file>>

== Running a sidechain client

=== Using a bash script
A script containing the `./bin/run start` command can be passed to PM2 to initiate the sidechain client.

Let's create a script  for xref:{url_build_hello_client}[The "Hello World" client].

> Create a `run_hello_client.sh` file inside the sidechain client's root directory and paste the following command in it.

Author: ```suggestion
Create a `run_hello_client.sh` file inside the sidechain client's root directory and then copy and paste the following command into it.
```

[source,json]
----

> ./bin/run start --config=config/custom_config.json

Reviewer: have you tested this, does it require an absolute path?

Author: Yes, it works. I created the `.sh` file within the sidechain client's root directory so I didn't need to mention the absolute path.

[source,bash]
----
pm2 start run_hello_client.sh
----

.Response
----
┌------------------------------------------------------------------------------------------------------------------------------------------------┐
│ id  │ name             │ namespace   │ version │ mode    │ pid      │ uptime │ ↺    │ status    │ cpu    │ mem    │ user   │ watching  │
├------------------------------------------------------------------------------------------------------------------------------------------------┤
│ 0   │ run_hello_client │ default     │ 0.1.0   │ fork    │ 55569    │ 21s    │ 30   │ online    │ 0%     │ 1.7mb  │ XYZ    │ disabled  │
└------------------------------------------------------------------------------------------------------------------------------------------------┘
----

=== Using a JSON config file
For advanced node operators, PM2 also accepts a configuration file that can have a range of options for various sidechain clients.

Create a PM2 config in the sidechain client's root directory, for example as shown below:

.pm2.conf.json
[source,json]
----
{
  "apps": [
    {
      "name": "hello_client",

>       "script": "./bin/run start --config=config/custom_config.json",

Reviewer: have you tested this one if this works? here script refers to the script name present in the package.json file of a node project.

Author: Yup, this one works as well, as you can see in its output, the PM2 shows the `hello_client` in the name column of its log message.

I am not sure about the second point as when I wrote the mentioned code in the script property it ran as usual.

      "instances": 1,
    }
  ]
}
----

> TIP: All available options for `script` and other PM2 commands can be found in `pm2 --help`. For more information on using JSON files for PM2, see {url_PM2_article}[PM2 — Advanced App Configuration with JSON File].

Reviewer 2: ```suggestion
TIP: All available options for `script` and other PM2 commands can be found in `pm2 --help`. For more information on using JSON files for PM2, see the {url_PM2_article}[PM2 — Advanced App Configuration with JSON File].
```

To add the process to PM2 and start it directly, execute the following command:

[source,bash]
----
pm2 start pm2.conf.json
----

.Response
----
┌------------------------------------------------------------------------------------------------------------------------------------------------┐
│ id  │ name             │ namespace   │ version │ mode    │ pid      │ uptime │ ↺    │ status    │ cpu    │ mem     │ user  │ watching  │
├------------------------------------------------------------------------------------------------------------------------------------------------┤
│ 0   │ hello_client     │ default     │ N/A     │ fork    │ 55594    │ 0 s    │ 0    │ online    │ 0%     │ 640.0kb │ XYZ   │ disabled  │
└------------------------------------------------------------------------------------------------------------------------------------------------┘
----

> TIP: You can create the *`pm2.conf.json`* or *`run_hello_client.sh`* files anywhere on the disk and can point to the *`./bin/run start`* command using an absolute path.

Reviewer 2: ```suggestion
TIP: You can create the *`pm2.conf.json`* or *`run_hello_client.sh`* files anywhere on the disk and point it to the *`./bin/run start`* command using an absolute path.
```

== Various PM2 commands

> A few important PM2 commands are discussed ahead, for more information, please refer to {url_pm2_docs}[PM2 docs^].

Reviewer 2: ```suggestion
A few important PM2 commands are covered in this section, for more information please refer to {url_pm2_docs}[PM2 docs^].
```
